zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/Zab1_0Test.java

```diff
@@ -701,6 +712,309 @@ private void proposeSetData(QuorumPacket qp, long zxid, String data, int version
+            readPacketSkippingPing(ia, qp);
+            assertEquals(Leader.ACK, qp.getType());
+            assertEquals(createZxid0, qp.getZxid());
+
+            // Read the UPTODATE ack.
+            readPacketSkippingPing(ia, qp);
+            assertEquals(Leader.ACK, qp.getType());
+            assertEquals(ZxidUtils.makeZxid(1, 0), qp.getZxid());
+
+            // The follower now starts following the leader.
+            // We send a PROPOSAL and a COMMIT, and wait for the transaction to be flushed by SyncRequestProcessor.
+            blocker = ((BlockingFollowerZooKeeperServer) follower.zk).blocker;
+            blocker.phaser.register();
+            long createZxid1 = ZxidUtils.makeZxid(1, 3);
+            qp.setType(Leader.PROPOSAL);
+            qp.setZxid(createZxid1);
+            hdr = new TxnHeader(13, 1313, createZxid1, 33, OpCode.create);
+            ct = new CreateTxn("/bar", "hi".getBytes(), Ids.OPEN_ACL_UNSAFE, false, 1);
+            baos = new ByteArrayOutputStream();
+            boa = BinaryOutputArchive.getArchive(baos);
+            boa.writeRecord(hdr, null);
+            boa.writeRecord(ct, null);
+            qp.setData(baos.toByteArray());
+            oa.writeRecord(qp, null);
+
+            qp.setType(Leader.COMMIT);
+            qp.setZxid(createZxid1);
+            oa.writeRecord(qp, null);
+
+            // Wait for "fsync" to begin.
+            assertTrue(followerSetUp.await(10, TimeUnit.SECONDS));
+            blocker.phaser.arriveAndAwaitAdvance();
+
+            // Now we send another PROPOSAL and COMMIT, and wait for them to be applied to the data tree.
+            // They will not be attempted flushed yet, because the ongoing "fsync" is slow (waiting on the phaser).
+            long createZxid2 = ZxidUtils.makeZxid(1, 4);
+            qp.setType(Leader.PROPOSAL);
+            qp.setZxid(createZxid2);
+            hdr = new TxnHeader(13, 1314, createZxid2, 34, OpCode.create);
+            ct = new CreateTxn("/baz", "bye".getBytes(), Ids.OPEN_ACL_UNSAFE, false, 1);
+            baos = new ByteArrayOutputStream();
+            boa = BinaryOutputArchive.getArchive(baos);
+            boa.writeRecord(hdr, null);
+            boa.writeRecord(ct, null);
+            qp.setData(baos.toByteArray());
+            oa.writeRecord(qp, null);
+
+            qp.setType(Leader.COMMIT);
+            qp.setZxid(createZxid2);
+            oa.writeRecord(qp, null);
+
+            // Wait for the follower to observe the COMMIT, and apply the PROPOSAL to its data tree. Unfortunately,
+            // there's nothing to do but sleep here, as watches are triggered before the last processed id is updated.
+            long doom = System.currentTimeMillis() + 1000;
+            while (createZxid1 != follower.fzk.getLastProcessedZxid() && System.currentTimeMillis() < doom) {
+                Thread.sleep(1);
+            }
+            assertEquals(createZxid1, follower.fzk.getLastProcessedZxid());
```

Comment:
Running the unit test shows that this assertion is not always true.
After txn(1, counter=3) is flushed, SyncRequestProcessor can take() txn(1, 4) and add to the `toFlush` queue without a flush(), then poll() returns a null, and the processor flushes. Txn(1,4) may or may not flush; it depends on the order.
A simple workaround is to enable flushDelay (via zookeeper.flushDelay) so that a flush for txn(1,4) is not called from an incoming null request. Add a barrier maybe?
It's likely to happen in an earlier version of the JDK(e.g., 1.8, 10, 11, etc., it has yet to occur at JDK18 in my test env, but why?)

Reply:
Ah, thanks for spotting this. The intention was to wait for the _second_ create txn ID, not the first. 

The read variable is volatile, so changes should be visible, given enough (sleep) time to let the other thread do its work. 
The construct was copied from a different test in the same class (lines 1101–1107), so if this is still unstable, I'd expect that test to also need an update.